### PR TITLE
 Fix nodes power state on Azure

### LINKED ISSUE #68916: Azure Cloud Provider: Wrong implementation of `InstanceShutdownByProviderID`
/kind bug

**What happened**:
Azure cloud provider uses `VM.ProvisioningState` to determine if the VM is running. Which never returns true even if the VM is not running (ref: https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_vmss.go#L137). The provider should uses (and parse) InstanceView.Statuses array -- specifically `PowerState/running` status. This currently impacts taints (i.e this taint `node.cloudprovider.kubernetes.io/shutdown:NoSchedule` is never added to the node as it goes down)

Fix: 
1. Add new functioning example `GetVMPowerStatus` returns true/false
2. replace https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_instances.go#L145 logic to call this new function

@feiskyer

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

#68033 implements InstanceShutdownByProviderID for azure, but it uses `VM.ProvisioningState` to determine if the VM is running. Which never returns true even if the VM is not running.

The provider should uses (and parse) `InstanceView.Statuses` array -- specifically `PowerState/running` status. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68916

**Special notes for your reviewer**:

**Release note**:

```release-note
 Fix Azure nodes power state for InstanceShutdownByProviderID()
```

/sig azure
/priority critical-urgent
/milestone v1.12
/kind bug

/cc @khenidak  @andyzhangx @justaugustus

### CODE REVIEW

File: pkg/cloudprovider/providers/azure/azure_vmss.go [hunk 1]
```diff
@@ -128,13 +128,24 @@ func (ss *scaleSet) getVmssVM(nodeName string) (ssName, instanceID string, vm co
...
+// GetPowerStatusByNodeName returns the power state of the specified node.
+func (ss *scaleSet) GetPowerStatusByNodeName(name string) (powerState string, err error) {
 	_, _, vm, err := ss.getVmssVM(name)
 	if err != nil {
-		return provisioningState, err
+		return powerState, err
 	}
 
-	return *vm.ProvisioningState, nil
+	if vm.InstanceView != nil && vm.InstanceView.Statuses != nil {
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_vmss.go]: use one function for both azure_standard.go and azure_vmss?
```
	if vm.InstanceView != nil && vm.InstanceView.Statuses != nil {
 		statuses := *vm.InstanceView.Statuses
 		for _, status := range statuses {
 			state := to.String(status.Code)
 			if strings.HasPrefix(state, vmPowerStatePrefix) {
 				return strings.TrimPrefix(state, vmPowerStatePrefix), nil
 			}
 		}
 	}


```

File: pkg/cloudprovider/providers/azure/azure_vmss_cache.go [hunk 2]
```diff
@@ -199,6 +199,24 @@ func (ss *scaleSet) newVmssVMCache() (*timedCache, error) {
 			return nil, nil
 		}
 
+		// Get instanceView for vmssVM.
+		if result.InstanceView == nil {
+			viewCtx, viewCancel := getContextWithCancel()
+			defer viewCancel()
+			view, err := ss.VirtualMachineScaleSetVMsClient.GetInstanceView(viewCtx, resourceGroup, ssName, instanceID)
+			// There is possible the vmssVM gets removed just before this call.
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_vmss_cache.go]: It is possible that vmssVM gets removed just before this call.

File: pkg/cloudprovider/providers/azure/azure_vmss.go (hunk 1 above)
Comment (feiskyer) [reply pkg/cloudprovider/providers/azure/azure_vmss.go]: Much more complicated because they are different GO struct.

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**